walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,155 @@
+        // if not - return only slots on which mandatory attendees are free
+        if (withOptionalAttendees.isEmpty()) {
+            return getMeetingSlots(events, request.getAttendees(), request.getDuration());
+        } else {
+            return withOptionalAttendees;
+        }
+    }
+
+    /**
+     * Function to get all possible slots for a meeting of specified group of people,
+     * with no optional attendees.
+     */
+    Collection<TimeRange> getMeetingSlots(Collection<Event> events, Collection<String> attendees, long duration) {
+        // make an array for getEmptyTimeRanges function
+        // put there all significant events' START ad END points assuming that event is a segment on time line.
+        ArrayList<Point> meetingsPoints = new ArrayList<>();
+
+        for (Event event : events) {
+            if (doIntersect(event.getAttendees(), attendees)) {
+                meetingsPoints.add(new Point(Point.Type.START, event));
+                meetingsPoints.add(new Point(Point.Type.END, event));
+            }
+        }
+        
+        return getEmptyTimeRanges(meetingsPoints, TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, duration);
+    }
+
+    /**
+     * @return all segments from firstPoint to lastPoint inclusive which are no shorter than minLength
+     *.        and do not intersect with segments from intersectingSegments
+     * @param intersectingSegments - array of Points. To put segment in this array, you should put its
+     *.                              START and END Points there.
+     * @param firstPoint - beginning of all time. Returned collection will not contain time before it.
+     * @param lastPoint - end of all time. Returned collection will not contain time after it.
+     * @param minLength - minimal length for each of the returned intervals
+     * Algorithm:
+     *.        All empty slots can start only where some segment ends and end only where some segment starts (or at
+               firstPoint or lastPoint)
+     *.        There cannot be any other points from intersectingSegments inside the empty slot - only its start end end.
+     *         Sort all points in intersectingSegments in ascending order and look at them one by one.
+     *.        Count how many segments are currently covering the point you are looking at: get a counter
+     *.        which increments every time you look at START point and decrements every time you look at END point.
+     *.        If the point is covered with 0 segments - it is a start of segment which can be one of the answer
+     *.        segments if it has appropriate length.
+     *.        So every time we look at a point, we check if the previous interval was segment-free and if so, we check
+     *.        the length of that interval to be at leash minLength.
+     */
+    private Collection<TimeRange> getEmptyTimeRanges(ArrayList<Point> intersectingSegments, int firstPoint, int lastPoint, long minLength) {
```

Comment:
I now have only one Point class constructor, which takes Event object as one of parameters, so in order to put firstPoint and lastPoint to the segmentPoints array I need to create a new constructor or a dummy Event object. Also in this case I need a comment why I check for openSegments == 1 instead of openSegments == 0. I think that it is probably easier and more explicit to consider firstPoint and lastPoint in particular.

Reply:
Well, you could create another constructor that takes the type and the time of the point so you could add a dummy START point at the end of the day and an END point at the beginning of the day. The == 0 check would be valid, but you would still have to initialize openSegments with 1.
But yea, this may be too complicated for this purpose, so probably considering them in particular is fine.